### PR TITLE
fixes for #3209 and #3208 and #3008

### LINKED ISSUE #3008: EncMapFromEncoding calls malloc and memset with zero size
On Gentoo Linux, users reported a linkage failure with GCC 4.9 and glibc 2.22.

https://bugs.gentoo.org/show_bug.cgi?id=561938
```
.libs/libfontforge_la-encoding.o: In function `EncMapFromEncoding':
encoding.c:(.text+0x6c3d): undefined reference to `__warn_memset_zero_len'
collect2: error: ld returned 1 exit status
```
This is apparently triggered by a sanity check that glibc implements to ensure that code does not call `memset` with a zero length parameter.
```
In file included from /usr/include/string.h:635:0,
                 from ../lib/string.h:41,
                 from fontforge.h:35,
                 from fontforgevw.h:30,
                 from encoding.c:28:
In function ‘memset’,
    inlined from ‘EncMapFromEncoding’ at encoding.c:1868:11:
/usr/include/bits/string3.h:86:30: warning: call to ‘__warn_memset_zero_len’ declared with attribute warning: memset used with constant zero length parameter; this could be due to transposed parameters [enabled by default]
       __warn_memset_zero_len ();
                              ^
```
The code in question is this:
```
    base = enc->char_cnt;
    if ( enc->is_original )
	base = 0;
    else if ( enc->char_cnt<=256 )
	base = 256;
    else if ( enc->char_cnt<=0x10000 )
	base = 0x10000;
    encoded = malloc(base*sizeof(int32));
    memset(encoded,-1,base*sizeof(int32));
```
If `base` gets set to zero, malloc and memset will be called with zero size/length. The behavior of such code is not well-defined; malloc may return NULL or a pointer to a zero-length section of memory. Passing NULL to memset is not well-defined either.

### LINKED ISSUE #3208: stack-buffer-overflow in readttfencodings
Hi.

I found a stack-buffer-overflow bug in fontforge.

Please confirm.

Thanks.

PoC download : [PoC](https://github.com/gy741/PoC/raw/master/readttfencodings)

```
=================================================================
==13820==ERROR: AddressSanitizer: stack-buffer-overflow on address 0x7fff76acb7c0 at pc 0x7f14c8acabe4 bp 0x7fff76acb350 sp 0x7fff76acb348
WRITE of size 2 at 0x7fff76acb7c0 thread T0
    #0 0x7f14c8acabe3 in readttfencodings /root/karas/fontforge/fontforge/parsettf.c:4729:12
    #1 0x7f14c8aa0428 in readttf /root/karas/fontforge/fontforge/parsettf.c:5588:2
    #2 0x7f14c8a8ff6b in _SFReadTTF /root/karas/fontforge/fontforge/parsettf.c:6291:11
    #3 0x7f14c8ed1511 in _ReadSplineFont /root/karas/fontforge/fontforge/splinefont.c:1115:11
    #4 0x7f14c8ed4d16 in ReadSplineFont /root/karas/fontforge/fontforge/splinefont.c:1286:9
    #5 0x7f14c8ed4d16 in LoadSplineFont /root/karas/fontforge/fontforge/splinefont.c:1344
    #6 0x7f14c8c8a05f in bOpen /root/karas/fontforge/fontforge/scripting.c:1858:10
    #7 0x7f14c8d18dda in docall /root/karas/fontforge/fontforge/scripting.c:9596:6
    #8 0x7f14c8d0dd4e in handlename /root/karas/fontforge/fontforge/scripting.c:9707:2
    #9 0x7f14c8d0dd4e in term /root/karas/fontforge/fontforge/scripting.c:9951
    #10 0x7f14c8d0c35d in mul /root/karas/fontforge/fontforge/scripting.c:10096:5
    #11 0x7f14c8d0a445 in add /root/karas/fontforge/fontforge/scripting.c:10141:5
    #12 0x7f14c8d093fb in comp /root/karas/fontforge/fontforge/scripting.c:10216:5
    #13 0x7f14c8d08268 in _and /root/karas/fontforge/fontforge/scripting.c:10259:5
    #14 0x7f14c8d055db in _or /root/karas/fontforge/fontforge/scripting.c:10290:5
    #15 0x7f14c8d055db in assign /root/karas/fontforge/fontforge/scripting.c:10322
    #16 0x7f14c8c6fa52 in expr /root/karas/fontforge/fontforge/scripting.c:10400:5
    #17 0x7f14c8c6fa52 in ff_statement /root/karas/fontforge/fontforge/scripting.c:10609
    #18 0x7f14c8c743fa in ProcessNativeScript /root/karas/fontforge/fontforge/scripting.c:10757:4
    #19 0x7f14c8c7672a in _CheckIsScript /root/karas/fontforge/fontforge/scripting.c:10849:3
    #20 0x7f14c8c7672a in CheckIsScript /root/karas/fontforge/fontforge/scripting.c:10886
    #21 0x7f14cac16e55 in fontforge_main /root/karas/fontforge/fontforgeexe/startui.c:1058:5
    #22 0x7f14c691c82f in __libc_start_main /build/glibc-bfm8X4/glibc-2.23/csu/../csu/libc-start.c:291
    #23 0x418c18 in _start (/root/karas/fontforge/fontforgeexe/.libs/fontforge+0x418c18)

Address 0x7fff76acb7c0 is located in stack of thread T0 at offset 1120 in frame
    #0 0x7f14c8abf86f in readttfencodings /root/karas/fontforge/fontforge/parsettf.c:4541

  This frame has 3 object(s):
    [32, 532) 'buffer.i'
    [608, 1120) 'table' <== Memory access at offset 1120 overflows this variable
    [1184, 1248) 'desired_cmaps'
HINT: this may be a false positive if your program uses some custom stack unwind mechanism or swapcontext
      (longjmp and C++ exceptions *are* supported)
SUMMARY: AddressSanitizer: stack-buffer-overflow /root/karas/fontforge/fontforge/parsettf.c:4729:12 in readttfencodings
Shadow bytes around the buggy address:
  0x10006ed516a0: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 04 f2
  0x10006ed516b0: f2 f2 f2 f2 f2 f2 f2 f2 00 00 00 00 00 00 00 00
  0x10006ed516c0: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x10006ed516d0: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x10006ed516e0: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
=>0x10006ed516f0: 00 00 00 00 00 00 00 00[f2]f2 f2 f2 f2 f2 f2 f2
  0x10006ed51700: 00 00 00 00 00 00 00 00 f3 f3 f3 f3 00 00 00 00
  0x10006ed51710: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x10006ed51720: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x10006ed51730: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x10006ed51740: 00 00 00 00 00 00 00 00 00 00 00 00 f1 f1 f1 f1
Shadow byte legend (one shadow byte represents 8 application bytes):
  Addressable:           00
  Partially addressable: 01 02 03 04 05 06 07
  Heap left redzone:       fa
  Heap right redzone:      fb
  Freed heap region:       fd
  Stack left redzone:      f1
  Stack mid redzone:       f2
  Stack right redzone:     f3
  Stack partial redzone:   f4
  Stack after return:      f5
  Stack use after scope:   f8
  Global redzone:          f9
  Global init order:       f6
  Poisoned by user:        f7
  Container overflow:      fc
  Array cookie:            ac
  Intra object redzone:    bb
  ASan internal:           fe
  Left alloca redzone:     ca
  Right alloca redzone:    cb
==13820==ABORTING
```

### LINKED ISSUE #3209: heap buffer overflow in readcffset
Hi.

I found a heap-buffer-overflow bug in fontforge.

Please confirm.

Thanks.

PoC download : [PoC](https://github.com/gy741/PoC/raw/master/readcffset)

```
=================================================================
==31382==ERROR: AddressSanitizer: heap-buffer-overflow on address 0x61600000d793 at pc 0x7fec747b75e3 bp 0x7ffc2c4fe230 sp 0x7ffc2c4fe228
WRITE of size 16 at 0x61600000d793 thread T0
    #0 0x7fec747b75e2 in readcffset /root/karas/fontforge/fontforge/parsettf.c:3555:26
    #1 0x7fec7478c64f in readcffglyphs /root/karas/fontforge/fontforge/parsettf.c:3971:2
    #2 0x7fec7477208f in readttf /root/karas/fontforge/fontforge/parsettf.c:5552:8
    #3 0x7fec74765f6b in _SFReadTTF /root/karas/fontforge/fontforge/parsettf.c:6291:11
    #4 0x7fec74ba7511 in _ReadSplineFont /root/karas/fontforge/fontforge/splinefont.c:1115:11
    #5 0x7fec74baad16 in ReadSplineFont /root/karas/fontforge/fontforge/splinefont.c:1286:9
    #6 0x7fec74baad16 in LoadSplineFont /root/karas/fontforge/fontforge/splinefont.c:1344
    #7 0x7fec7496005f in bOpen /root/karas/fontforge/fontforge/scripting.c:1858:10
    #8 0x7fec749eedda in docall /root/karas/fontforge/fontforge/scripting.c:9596:6
    #9 0x7fec749e3d4e in handlename /root/karas/fontforge/fontforge/scripting.c:9707:2
    #10 0x7fec749e3d4e in term /root/karas/fontforge/fontforge/scripting.c:9951
    #11 0x7fec749e235d in mul /root/karas/fontforge/fontforge/scripting.c:10096:5
    #12 0x7fec749e0445 in add /root/karas/fontforge/fontforge/scripting.c:10141:5
    #13 0x7fec749df3fb in comp /root/karas/fontforge/fontforge/scripting.c:10216:5
    #14 0x7fec749de268 in _and /root/karas/fontforge/fontforge/scripting.c:10259:5
    #15 0x7fec749db5db in _or /root/karas/fontforge/fontforge/scripting.c:10290:5
    #16 0x7fec749db5db in assign /root/karas/fontforge/fontforge/scripting.c:10322
    #17 0x7fec74945a52 in expr /root/karas/fontforge/fontforge/scripting.c:10400:5
    #18 0x7fec74945a52 in ff_statement /root/karas/fontforge/fontforge/scripting.c:10609
    #19 0x7fec7494a3fa in ProcessNativeScript /root/karas/fontforge/fontforge/scripting.c:10757:4
    #20 0x7fec7494c72a in _CheckIsScript /root/karas/fontforge/fontforge/scripting.c:10849:3
    #21 0x7fec7494c72a in CheckIsScript /root/karas/fontforge/fontforge/scripting.c:10886
    #22 0x7fec768ece55 in fontforge_main /root/karas/fontforge/fontforgeexe/startui.c:1058:5
    #23 0x7fec725f282f in __libc_start_main /build/glibc-bfm8X4/glibc-2.23/csu/../csu/libc-start.c:291
    #24 0x418c18 in _start (/root/karas/fontforge/fontforgeexe/.libs/fontforge+0x418c18)

0x61600000d793 is located 9 bytes to the right of 522-byte region [0x61600000d580,0x61600000d78a)
allocated by thread T0 here:
    #0 0x4b8d48 in __interceptor_malloc (/root/karas/fontforge/fontforgeexe/.libs/fontforge+0x4b8d48)
    #1 0x7fec747b2cb3 in readcffset /root/karas/fontforge/fontforge/parsettf.c:3536:18

SUMMARY: AddressSanitizer: heap-buffer-overflow /root/karas/fontforge/fontforge/parsettf.c:3555:26 in readcffset
Shadow bytes around the buggy address:
  0x0c2c7fff9aa0: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c2c7fff9ab0: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x0c2c7fff9ac0: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x0c2c7fff9ad0: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x0c2c7fff9ae0: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
=>0x0c2c7fff9af0: 00 02[fa]fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c2c7fff9b00: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c2c7fff9b10: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x0c2c7fff9b20: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x0c2c7fff9b30: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x0c2c7fff9b40: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
Shadow byte legend (one shadow byte represents 8 application bytes):
  Addressable:           00
  Partially addressable: 01 02 03 04 05 06 07
  Heap left redzone:       fa
  Heap right redzone:      fb
  Freed heap region:       fd
  Stack left redzone:      f1
  Stack mid redzone:       f2
  Stack right redzone:     f3
  Stack partial redzone:   f4
  Stack after return:      f5
  Stack use after scope:   f8
  Global redzone:          f9
  Global init order:       f6
  Poisoned by user:        f7
  Container overflow:      fc
  Array cookie:            ac
  Intra object redzone:    bb
  ASan internal:           fe
  Left alloca redzone:     ca
  Right alloca redzone:    cb
==31382==ABORTING
```

### PR DESCRIPTION
**Pull request by skef**
Avoid calling malloc or memcpy with 0, sanity check a `gid` read against the glyph count.  

Closes #3209 
Closes #3208 
Closes #3008

### CODE REVIEW

File: fontforge/encoding.c [hunk 1]
```diff
@@ -2101,14 +2105,22 @@ return( NULL );
...
-    memcpy(map->map,encoded,base*sizeof(int32));
-    memcpy(map->map+base,unencoded,extras*sizeof(int32));
+    if ( map->enccount>0 ) {
+	map->map = malloc(map->enccount*sizeof(int32));
+	if ( base>0 )
+	    memcpy(map->map,encoded,base*sizeof(int32));
+	if ( extras>0 )
+	    memcpy(map->map+base,unencoded,extras*sizeof(int32));
+    } else
+	map->map = NULL;
```
Comment (jtanx) [inline fontforge/encoding.c]: This should be unnecessary, given it's allocated with chunkalloc

Comment (skef) [reply fontforge/encoding.c]: I suppose, but it seems like skipping these often results in a compiler warning.

**Comment by skef** [comment]
Tests worked in debug, maybe they're broken in release? I'll check.

File: fontforge/encoding.c [hunk 2]
```diff
@@ -2031,10 +2031,14 @@ return( NULL );
 	base = 256;
     else if ( enc->char_cnt<=0x10000 )
 	base = 0x10000;
-    encoded = malloc(base*sizeof(int32));
-    memset(encoded,-1,base*sizeof(int32));
-    unencoded = malloc(sf->glyphcnt*sizeof(int32));
-    unmax = sf->glyphcnt;
+    if ( base>0 ) {
+	encoded = malloc(base*sizeof(int32));
```
Comment (jtanx) [inline fontforge/encoding.c]: Try null initializing encoded and unencoded on line 2019, these will be garbage if base/glyphcnt <= 0 and there's reallocs below.

File: fontforge/encoding.c (hunk 1 above)
Comment (jtanx) [reply fontforge/encoding.c]: I dont't think the compiler would warn in this case, it's more likely to warn on uninitialised stack variables.

File: fontforge/encoding.c (hunk 2 above)
Comment (skef) [reply fontforge/encoding.c]: That was it!

**Comment by skef** [comment]
(added missed fixed bug number #3008 to title and description)

File: fontforge/encoding.c (hunk 1 above)
Comment (skef) [reply fontforge/encoding.c]: You approved, but all things considered would you prefer I remove these?

Comment (jtanx) [reply fontforge/encoding.c]: Probably, but it's not something I'll lose sleep over

**Comment by ctrlcctrlv** [comment]
It might also fix some of the CID bugs I found at the end of last year. I don't know if it's worth leaving this open while I confirm or disconfirm&dagger; that, but it might be worth me checking after this is merged

<sup>&dagger; Maybe not a word.</sup>

**Comment by skef** [comment]
> I don't know if it's worth leaving this open while I confirm or disconfirm† that, but it might be worth me checking after this is merged

@ctrlcctrlv Is the idea that you might compile the branch prior to merging, to see if we should add some additional bugs to the description? I'm happy to wait on that, these aren't pressing. (Obviously it would be tricky to merge while leaving the PR open.)

**Comment by ctrlcctrlv** [comment]
Yes, that is my idea. I already have a remote named `skef` (and `jtanx`) so it's really not a problem for me to compile your branch as opposed to mine. I still have all the test cases from those CID map bugs.

**Comment by ctrlcctrlv** [comment]
I had the greatest reason to believe that this might fix #3948 or #3946, which happen even if the encoding is not compacted. Unfortunately, it fixes neither.

I had little hope it would solve the bug with compact encodings. I tested #3947 again and it doesn't fix that either.

So, sorry for the unnecessary delay - this can be merged now.

**Pull request merged**